ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+Once all the required workflow items are executed the workunit is marked as completed.  Alternatively, if there are
+workflow items which are waiting to be triggered by an event the workunit will be passed to the scheduler
+which will keep monitoring for events.
+
+There are various specialised types of workflow items - e.g., sequential, wait, independent, but they all follow the same basic approach of
+executing dependencies and then executing that particular item.  
+
+Most of the interesting work in a ECL query is done within a graph.  The call ctx->executeGraph will either execute the graph locally
+(in the case of hthor and roxie), or add the workunit onto a queue (for Thor).  Whichever happens, control will pass to that engine...
+
+Executing a graph
+=================
+
+All the engines (roxie, hThor, Thor) execute graphs in a very similar way.  The main differences are that hThor and
+Thor execute a sub graph at a time, while roxie executes a complete graph as one.  Roxie is also optimized to minimize
+the overhead of executing a query - since the same query tends to be run multiple times - which means that it creates a
+graph of factory objects and those are used to create the activities.  The core details are the same for each of them though.
+
+Details of the graph structure
+------------------------------
+
+A reminder of the structure of the graph with the xml for the graph definition in our example.  Each graph (e.g., graph1)
+consists of 1 or more subgraphs (e.g., node id=1).  Each of those subgraphs contains 1 or more activities (node id=2, node id=3).
+
+::
+
+  <Graph name="graph1" type="activities">
+   <xgmml>
+    <graph wfid="2">
+     <node id="1">
+      <att>
+       <graph>
+        <att name="rootGraph" value="1"/>
+        <edge id="2_0" source="2" target="3"/>
+        <node id="2" label="Index Read&#10;&apos;names&apos;">
+         <att name="definition" value="workuniteg1.ecl(3,1)"/>
+         <att name="name" value="results"/>
+         <att name="_kind" value="77"/>
+         <att name="ecl" value="INDEX({ string40 name, string80 address }, &apos;names&apos;, fileposition(false));&#10;FILTER(KEYED(name = STORED(&apos;searchname&apos;)));&#10;"/>
+         <att name="recordSize" value="120"/>
+         <att name="predictedCount" value="0..?[disk]"/>
+         <att name="_fileName" value="names"/>
+        </node>
+        <node id="3" label="Output&#10;Result #1">
+         <att name="definition" value="workuniteg1.ecl(4,1)"/>
+         <att name="_kind" value="16"/>
+         <att name="ecl" value="OUTPUT(..., workunit);&#10;"/>
+         <att name="recordSize" value="120"/>
+        </node>
+       </graph>
+      </att>
+     </node>
+    </graph>
+   </xgmml>
+  </Graph>
+  
+The xml for each activity might contain the following information:
+
+* A unique id (e.g. id="2").
+* The "kind" of the activity e.g., <att name="_kind" value="77"/>.  The value is an item from the enum ThorActivityKind in eclhelper.hpp.
```

Comment:
http://en.wiktionary.org/wiki/e.g
A punctuation mark always precedes e.g.
I like sweet foods, e.g. chocolate. (cf. I like sweet foods, for example chocolate.)
I like sweet foods (e.g. chocolate and marzipan) and eat them often.